src/measure_theory/ess_sup.lean

```diff
@@ -0,0 +1,119 @@
+`α →ₘ[μ] β` (see measure_theory/ae_eq_fun.lean).
+
+## Main definitions
+
+* `ess_sup f μ := μ.ae.limsup f`
+* `ess_inf f μ := μ.ae.liminf f`
+-/
+
+open measure_theory
+
+variables {α β : Type*} [measurable_space α] {μ : measure α}
+
+section conditionally_complete_lattice
+variable [conditionally_complete_lattice β]
+
+/-- Essential supremum of `f` with respect to measure `μ`: the smallest `c : β` such that
+`f x ≤ c` a.e. -/
+def ess_sup (f : α → β) (μ : measure α) := μ.ae.limsup f
+
+/-- Essential infimum of `f` with respect to measure `μ`: the greatest `c : β` such that
+`c ≤ f x` a.e. -/
+def ess_inf (f : α → β) (μ : measure α) := μ.ae.liminf f
+
+lemma ess_sup_congr_ae {f g : α → β} (hfg : f =ᵐ[μ] g) : ess_sup f μ = ess_sup g μ :=
+filter.limsup_congr hfg
+
+lemma ess_inf_congr_ae {f g : α → β} (hfg : f =ᵐ[μ] g) :  ess_inf f μ = ess_inf g μ :=
+@ess_sup_congr_ae α (order_dual β) _ _ _ _ _ hfg
+
+end conditionally_complete_lattice
+
+section complete_lattice
+variable [complete_lattice β]
+
+@[simp] lemma ess_sup_measure_zero {f : α → β} : ess_sup f 0 = ⊥ :=
+le_bot_iff.mp (Inf_le (by simp [set.mem_set_of_eq, filter.eventually_le, ae_iff]))
+
+@[simp] lemma ess_inf_measure_zero {f : α → β} : ess_inf f 0 = ⊤ :=
+@ess_sup_measure_zero α (order_dual β) _ _ _
+
+lemma ess_sup_mono_ae {f g : α → β} (hfg : f ≤ᵐ[μ] g) :  ess_sup f μ ≤ ess_sup g μ :=
+filter.limsup_le_limsup hfg
+
+lemma ess_inf_mono_ae {f g : α → β} (hfg : f ≤ᵐ[μ] g) : ess_inf f μ ≤ ess_inf g μ :=
+filter.liminf_le_liminf hfg
+
+lemma ess_sup_const (c : β) (hμ : μ ≠ 0) : ess_sup (λ x : α, c) μ = c :=
+begin
+  have hμ_ne_bot : μ.ae.ne_bot, by rwa [filter.ne_bot, ne.def, ae_eq_bot],
+  exact @filter.limsup_const _ _ _ μ.ae hμ_ne_bot c,
+end
+
+lemma ess_inf_const (c : β) (hμ : μ ≠ 0) : ess_inf (λ x : α, c) μ = c :=
+@ess_sup_const α (order_dual β) _ _ _ _ hμ
+
+lemma order_iso.ess_sup_apply {γ} [complete_lattice γ] (f : α → β) (μ : measure α)
+  (g : β ≃o γ) :
+  g (ess_sup f μ) = ess_sup (λ x, g (f x)) μ :=
+begin
+  refine order_iso.limsup_apply g _ _ _ _,
+  all_goals { by filter.is_bounded_default},
+end
+
+lemma order_iso.ess_inf_apply {γ} [complete_lattice γ] (f : α → β) (μ : measure α)
+  (g : β ≃o γ) :
+  g (ess_inf f μ) = ess_inf (λ x, g (f x)) μ :=
+@order_iso.ess_sup_apply α (order_dual β) _ _  (order_dual γ) _ _ _ g.dual
+
+end complete_lattice
+
+section complete_linear_order
+variable [complete_linear_order β]
+
+lemma ae_lt_of_ess_sup_lt {f : α → β} {x : β} (hf : ess_sup f μ < x) : ∀ᵐ y ∂μ, f y < x :=
+filter.eventually_lt_of_limsup_lt hf
+
+lemma ae_lt_of_lt_ess_inf {f : α → β} {x : β} (hf : x < ess_inf f μ) : ∀ᵐ y ∂μ, x < f y :=
+@ae_lt_of_ess_sup_lt α (order_dual β) _ _ _ _ _ hf
+
+end complete_linear_order
+
+section ennreal
+
+lemma ennreal.ae_le_ess_sup (f : α → ennreal) : ∀ᵐ y ∂μ, f y ≤ ess_sup f μ :=
+ennreal.eventually_le_limsup f
+
+lemma ennreal.ess_sup_const_mul {f : α → ennreal} {a : ennreal} (ha_top : a ≠ ⊤) :
```

Comment:
Isn't this one also true for `a = ⊤`?

Reply:
Yes it is and I changed it, but I am surprised at the amount of work that modification required.